### PR TITLE
server: fix race condition for heartbeat

### PR DESCRIPTION
**Pull request by R4PaSs**
The heartbeater works on a timer, and is run in a goroutine to keep
track of the state of a connection.

The problem is though that the timer is not closed when the connection
is, and therefore, it may try to write into the connection outgoing
channel, making the whole application panic.

This commit fixes such a behaviour by ensuring the connection is not
closed when checking for the heartbeat.

### CODE REVIEW

**Comment by valinurovam** [comment]
@R4PaSs thx for fix! But it will be better to set a ticker as connection property and Stop it inside connection.close before `conn.netConn.Close()`

**Comment by R4PaSs** [comment]
That makes sense, I'll reroll this soon!

**Comment by R4PaSs** [comment]
Hi @valinurovam, I've rerolled the PR, the timer is now stopped when the server shuts down, this should do the trick

File: server/connection.go [hunk 1]
```diff
@@ -119,6 +120,11 @@ func (conn *Connection) close() {
 		conn.statusLock.Unlock()
 		return
 	}
+
+	if conn.heartbeatTimer != nil {
```
Comment (valinurovam) [inline server/connection.go]: Not critical, but Lock - Unlock is `critical section` and should be much faster, can u move Stop after unlock?

**Comment by valinurovam** [comment]
@R4PaSs merged, thanks! The issue with lock-unlock section can be fixed later.

**Pull request merged**